### ISSUE TITLE
Consider replacing Webpack with Vite

### ISSUE THREAD

**Issue by silverwind** [go-gitea/gitea#17793]
### Feature Description

It might be a bit early, but I think it's worth investigating switching the build process to pure esbuild. It supports [experimental code splitting](https://esbuild.github.io/api/#splitting) and has [a plugin ecosystem](https://github.com/esbuild/community-plugins) as well. It should build supremely fast.

Vite was suggested previously as a Webpack replacment but I think it's a dead-end because it relies on legacy systems like Rollup and Babel and seems to be generally header into a similar bloated direction like Webpack. Also, in terms of speed, Vite is barely faster than Webpack in my testing.

**Comment by silverwind** [comment]
https://github.com/evanw/esbuild/issues/1441 could be one potential blocker for us, given that we modify the public path prefix at runtime.

**Comment by lafriks** [comment]
Can't we just build with public path `{{PUBLIC_PATH}}/` and then just replace it in gitea server code either on gitea startup or on first usage before serving?

**Comment by silverwind** [comment]
Yes there are ways. Probably just using the existing `window.config` mechanism would be the easiest.

**Comment by anbraten** [comment]
> Also, in terms of speed, Vite is barely faster than Webpack in my testing.

I thing this is mainly coming from the fact how the code is currently structured. Vites power is code-splitting and being fast on dynamically importing code on demand. In case of the Gitea code it seems to me that it loads the features a monolithic bundle and "only" uses dynamic imports for libraries atm, so Vite probably can't optimize that to well. 

If Vue is a thing for Gitea, Vite should be preferred over esbuild from my exp considering the setup complexity for vue SFC building with native esbuild. Maybe https://vitejs.dev/guide/why.html#why-not-bundle-with-esbuild has some helping hints as well.

**Comment by silverwind** [comment]
> Maybe https://vitejs.dev/guide/why.html#why-not-bundle-with-esbuild has some helping hints as well.

That page looks a bit outdated to me, esbuild can now do CSS minification and code splitting (althought beta-level). Vue SFCs will probably require a plugin.

**Comment by silverwind** [comment]
I guess Vite has matured enough to be a possible Webpack replacement. It will require quite a bit of configuration and probably some plugins, but I think it could be made to work.

I see Vite currently a possibly more "complete" version of esbuild, which still lacks [code splitting](https://github.com/evanw/esbuild/issues/16).

**Comment by silverwind** [comment]
Actually, there are at least two blockers I think:

- Vite can't do multiple CSS entrypoints. The entrypoint config is provided by rollup and I think rollup can only do JS entrypoints.
- `license.txt` generation will need to be rewritten or a suitable plugin to be found.

**Comment by silverwind** [comment]
I don't see a compelling reason to switch to vite currently for Gitea. Our webpack setup is supremely fast currently at around 20s to build, and that includes heavyweights like monaco. Vite just does not seem to be made for complex projects that need multiple entry points and fine-grained configurability.

**Comment by silverwind** [comment]
Let's reopen. Vite is still a goal and I think with some Rollup plugins to load CSS, we may be able to achieve what Webpack currently does.

**Comment by silverwind** [comment]
Likely use https://github.com/mjeanroy/rollup-plugin-license for `license.txt` generation with vite.

**Comment by silverwind** [comment]
Attempt to migrate to Vite was done in https://github.com/go-gitea/gitea/pull/25874, but it did not work out:

- Vite, or more specifically Rollup is 2-3 times slower to bundle our dependencies, likely needs https://github.com/evanw/esbuild/issues/16 and adoption of esbuild bundling in vite to really speed it up
- Vite is heavily geared towards SPAs, or at least apps where it can serve HTML. This will never be the case for gitea, so it's a bad fit and would require a ton of workarounds to get working.

**Comment by silverwind** [comment]
Vite should be re-attempted again after they release a version based on [rolldown](https://github.com/rolldown/rolldown), which should provide a level of speed suitable to us.

**Comment by silverwind** [comment]
Maybe https://github.com/farm-fe/farm is worth a try in the future. It promises to alleviate all problems I had with vite (rollup slowness, code splitting and slow dev server because of nonexistant lazy-compilation).
